drivers/flickr/onitu_flickr/flickr_driver.py

```diff
@@ -0,0 +1,383 @@
+        raise ServiceError(u"Failed to delete photo {}: error {} {}"
+                           .format(metadata.filename, fe.code, fe.message))
+    plug.logger.debug(u"Deleted photo {} (ID {})"
+                      .format(metadata.filename, photoID))
+
+
+@plug.handler()
+def move_file(old_metadata, new_metadata):
+    plug.logger.debug(u"Moving photo {} to {}"
+                      .format(old_metadata.filename, new_metadata.filename))
+    photoID = old_metadata.extra['photo_id']
+    if photoID is None:
+        raise DriverError(u"No Photo ID for {} !"
+                          .format(old_metadata.filename))
+    # Respecting the Flickr convention that wants that photo titles be
+    # without file extension.
+    title = os.path.splitext(new_metadata.filename)[0]
+    try:
+        flickr.flickr_api.photos.setMeta(photo_id=photoID, title=title)
+    except FlickrError as fe:
+        raise ServiceError(u"Failed to move file {} to {} on Flickr: {}"
+                           .format(old_metadata.filename,
+                                   new_metadata.filename, fe.message))
+    new_metadata.extra['photo_id'] = photoID
+    now = int((datetime.utcnow() - datetime(1970, 1, 1)).total_seconds())
+    new_metadata.extra['last_update'] = now
+    new_metadata.write()
+    plug.logger.debug(u"Successfully moved {} to {}"
+                      .format(old_metadata.filename, new_metadata.filename))
+
+
+class CheckChanges(threading.Thread):
+    """A class spawned in a thread to poll for changes on the Dropbox account.
+    Dropbox works with deltas so we have to periodically send a request to
+    Dropbox with our current delta to know what has changed since we retrieved
+    it."""
+
+    def __init__(self, folder, timer):
+        threading.Thread.__init__(self)
+        self.stopEvent = threading.Event()
+        self.folder = folder
+        self.timer = timer
+        self.photosetID = flickr.getPhotosetID(self.folder.path)
+        # Set default as 1 for the last update timestamp because Flickr throws
+        # an error if we start from 0... -.-
+        self.lastUpdateTimestamp = plug.service_db.get('timestamp', default=1)
+        plug.logger.debug("Getting timestamp {} out of database"
+                          .format(self.lastUpdateTimestamp))
+
+    def run(self):
+        while not self.stopEvent.isSet():
+            try:
+                # Getting it every time because of the photoset ID
+                # modifications that could occur during runtime (album
+                # automatic deletions, etc.)
+                self.photosetID = flickr.getPhotosetID(self.folder.path)
+                self.checkPhotoset()
+            except FlickrError as fe:
+                warn = u"An unknown Flickr error occurred: {}".format(
+                       fe.message)
+                if fe.message == u"Error: 1: Photoset not found":
+                    warn = (u"The album {} doesn't exist on Flickr."
```

Comment:
No, because in order to create a Flickr album we MUST set a primary photo to put in this album. The check changes thread isn't really able to do that...

I thought it's a 'not-so-abnormal' situation since e.g. Flickr deletes an album when you delete all photos in it (and it could happen when doing so from Onitu). So I think if the user wants to clean up a bit, it can be a normal situation that the album's temporarily gone.

The upload_file handler manages this situation. If the album we want to upload a photo in doesn't exist (or doesn't exist *anymore*), we (re-)create it.